### PR TITLE
ENH: Adapt LBFGS2Optimizerv4 to inherit from GradientDescentOptimizerV4

### PR DESCRIPTION
**Pull request by samuelgerber**
Add support for local transforms.

Inherit from GradientDescentOptimizerv4 to add support for threaded scaling and fit it into the hierarchy of gradient descent based optimizers.

### CODE REVIEW

File: Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx [hunk 1]
```diff
@@ -274,6 +293,7 @@ LBFGS2Optimizerv4::GetStopConditionDescription() const
 }
 
 // Disallow setting scale and weight
+/*
```
Comment (dzenanz) [inline Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx:296]: Why not remove this now?

**Review by dzenanz — COMMENTED** [review]
LGTM, but someone else should review too.

File: Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx (hunk 1 above)
Comment (samuelgerber) [reply Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx:296]: Yes should be removed.

File: Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx [hunk 2]
```diff
@@ -139,17 +141,29 @@ LBFGS2Optimizerv4::EvaluateCost(const LBFGS2Optimizerv4::PrecisionType * x,
                                 const LBFGS2Optimizerv4::PrecisionType)
 {
   ParametersType xItk(n);
-  std::memcpy(xItk.data_block(), x, n * sizeof(LBFGS2Optimizerv4::PrecisionType));
+  // TODO: check if this works with SSE allocation
+  // might need to copy data
+  xItk.SetData(const_cast<double *>(x), n, false);
```
Comment (samuelgerber) [inline Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx:146]: This is potentially dangerous but would avoid copying the data multiple times.
Is there are reason the SetParameters methods in the v4 metrics is not taking a const ParameterType?

File: Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx (hunk 1 above)
Comment (samuelgerber) [reply Modules/Numerics/Optimizersv4/src/itkLBFGS2Optimizerv4.cxx]: Removed in commit
https://github.com/InsightSoftwareConsortium/ITK/pull/2372/commits/0431e2297516b43f5f83e2100bc11b04dd420a65

**Comment by dzenanz** [comment]
LGTM, but I would like someone else more familiar with optimizers to review this.

@samuelgerber squashing the commits and removing WIP should entice someone else to look at this PR too.

**Review by dzenanz — APPROVED** [review]

**Comment by dzenanz** [comment]
Can someone else review too?

**Comment by samuelgerber** [comment]
I think this needs more testing on real world examples. (Maybe adapt some of the registration tests to use this optimizer as well).

**Comment by dzenanz** [comment]
Merged before this PR rots away.

**Pull request merged**